### ISSUE TITLE
Extension disabled due to new permissions

### ISSUE THREAD

**Issue by corollari** [bitwarden/browser#1548]
## Describe the Bug
This is not strictly a bug.

Since the latest update the browser has automatically disabled the extension due to the addition of a new permission needed to communicate with native applications. This is not needed for the extension to operate and thus could be made an optional permission that is only asked for when the user enables a feature that requires it.

Feature request associated with it: https://community.bitwarden.com/t/make-exchange-messages-with-programs-an-optional-permission/16893

## Steps To Reproduce
1. Install second to last version of bitwarden on chrome
2. Update version of bitwarden

## Expected Result
Nothing happens, when a user wants to use a feature that requires the new permission a popup appears requesting this permission.

## Actual Result
Extension is disabled, browser requests for new permission.

## Environment

- Operating system: any
- Browser: any

**Comment by clecap** [comment]
Sorry, but this **is** a bug.

Bitwarden is a highly sensitive security application managing 100 and more passwords. It is **not** a good idea to have this application require additional permissions to communicate with other applications. I rather take this as a worrying indication that the development of Bitwarden is turning into a bad and sad and wrong direction.

And, yes, Bitwarden should definitely make this additional request for permissions optional.

Where can I download the old version of the extension? I do **not** want this extension to operate with more permissions than is necessary for the most fundamental options.

**Comment by eight04** [comment]
> Where can I download the old version of the extension?

For Firefox, there is a version history page:
https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/versions/

**Comment by github-account1111** [comment]
@clecap I agree with the premise, but if security is important, then using older versions is counterproductive, as it leads to a potentially less secure environment than with an up-to-date version (even one that has more permissions).

**Comment by corollari** [comment]
I agree with @clecap, it would be nice for a security-focused software such as bitwarden to follow a least-privilege-needed approach.

**Comment by corollari** [comment]
Behaviour was introduced in commit 296ccb68298e7f1de591a16adce59b99cc40152a

**Comment by bialad** [comment]
Went digging around for release notes, and found this.

https://bitwarden.com/help/article/biometrics/#browser-extensions

So the change seems legit, but should for sure be optional. I won't be using this functionality for example.

**Comment by escape0707** [comment]
If we are going to implement optional permissions, maybe fix issue #883, too.

**Comment by escape0707** [comment]
> 
> 
> > Where can I download the old version of the extension?
> 
> For Firefox, there is a version history page:
> https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/versions/

Yet another reason to keep using Firefox rather than chromium. Yay!

**Comment by SergeantConfused** [comment]
Hi there,

This new permission allows the browser extension to communicate with your local desktop client to use biometrics to unlock the vault. More on this here: https://bitwarden.com/help/article/releasenotes/ and here: https://bitwarden.com/help/article/biometrics/#browser-extensions

I hope this clarifies this, please let me know if you have any further questions.

All the best,

**Comment by orrc** [comment]
@SergeantConfused Thanks. I think everybody here is aware of _why_ the permission was added, but it should be ideally have been made optional.

However, I think a major point is that this change was poorly communicated and documented to end users.

Thanks for letting us know about the [recent release notes change](https://github.com/bitwarden/help/pull/369) to mention this permission, but I [still think](https://twitter.com/orrc/status/1353027074446024704) that it would be great to make it a *lot* easier for users to find this information.

To that end I've opened a further docs PR: https://github.com/bitwarden/help/pull/370 🙂

**Comment by erasma** [comment]
Please make this browser change optional

**Comment by clecap** [comment]
@github-account1111 absolutely yes - provided the updates move into the right direction. Here I have, sorry to say, some serious doubts. While I certainly understand the convenience of all kinds of additional UI features and while I am certainly grateful that they exist they (1) definitely should be optional, (2) trade convenience for security, (3) were not reasonably communicated to end users and (4) came as a "oops, my system has been hacked" surprise to me.

And therefore my trust that updates move into the right direction of more secure software is, here, shaken. 

All I want from a password store is to keep my passwords safe - and communicating them to "cooperating programs" by means of some "click ok or have your password store disabled" is the textbook example of what I am not expecting from secure system design. Sorry.

**Comment by charis-the-programmer** [comment]
> Sorry, but this **is** a bug.
> 
> Bitwarden is a highly sensitive security application managing 100 and more passwords. It is **not** a good idea to have this application require additional permissions to communicate with other applications. I rather take this as a worrying indication that the development of Bitwarden is turning into a bad and sad and wrong direction.
> 
> And, yes, Bitwarden should definitely make this additional request for permissions optional.
> 
> Where can I download the old version of the extension? I do **not** want this extension to operate with more permissions than is necessary for the most fundamental options.

I agree with @clecap, I'm here because I thought my system was compromised. The ironic thing is I couldn't log in to comment on this thread because bitwarden was disabled, smh.

**Comment by Roebie** [comment]
This is absolutely a bug.
Looking at the [Unlock with Biometrics article](https://bitwarden.com/help/article/biometrics/#browser-extensions) I can tell that my computer doesn't meet the requirements to use this feature, that I won't be installing the desktop app at any time soon and that thus this additional permission is totally unneeded and unwanted.
So Bitwarden, get you act together and make this optional. My only alternative would be to find yet another password manager.

**Comment by kspearrin** [comment]
We're looking into a way to make the permission optional. [It is documented](https://extensionworkshop.com/documentation/develop/request-the-right-permissions/#request-permissions-at-runtime) that it is not possible to make `nativeMessaging` an optional permission with web extension APIs (which is originally why it was released like this), however, upon further trial, it appears we still can in Chromium browsers. However, [Firefox still does not allow it to be optional](https://bugzilla.mozilla.org/show_bug.cgi?id=1630415) so it would have to be left global there.

**Comment by Belphemur** [comment]
To add to this, don't forget that Bitwarden is also used by non-technical people. My partner wondered where did the extension go because chrome directly blocked the update, and she didn't feel comfortable giving "weird" permissions, her words, not mine.

As devs, we always have to remember that our software are used by non-technical people that will be scared by the smallest popup especially when it's an app that stores your passwords.

**Comment by escape0707** [comment]
Thanks for dev's hardwork! And guys, vote here https://bugzilla.mozilla.org/show_bug.cgi?id=1630415

**Comment by REMjn832** [comment]
Excited for a fix to this. It keeps disabling in Chromium based browsers as well. I will enable it, and 10 minutes later Chrome will disable it. Very annoying... Thanks!

**Comment by jaredm4** [comment]
I understand this is a permission that couldn't be set as optional in Firefox. What baffles me is that the Bitwarden team discussed this: Come Monday morning the **majority of users** are going to start their week with an un-usable password manager, demanding access to a questionable permission, so a **minority of users** will get their biometric unlocking. Not saying it's a bad feature, but when you weigh the pros/cons.... it shocks me this was let through.

On top of that, details of this change are hard to find, and even for myself THIS pull request was the first real search result mentioning it. I honestly was starting to wonder if the Bitwarden extension was compromised.

**Comment by dhruv** [comment]
@kspearrin We are all grateful for your work and the great product your team makes. However, this decision is quite disappointing. It leaves Bitwarden/Firefox users in a position where they have to do one of:

- Accept this new permission (bad OpSec)
- Give up Firefox and accept that they have to use a chromium-based browser
- Give up Bitwarden and move to a different password manager

I would argue the importance of those choices are OpSec > Browser > Password Manager.

Is forcing this choice for Firefox users the final position BitWarden devs would like to take? As a Firefox user, I'd like to petition you to remove this feature for now and let's request the Firefox devs to make this permission optional. Happy to help however I can.

**Comment by tanrr** [comment]
I cannot allow the Chrome Bitwarden browser extension this sort of external permission.

I will stop using the Bitwarden extension and recommend to other users that they stop using Bitwarden as well, until this permission is made optional (or simply not implemented for Firefox).

Please consider how many customers you are going to lose if you don't change your approach to this.

**Comment by cscharf** [comment]
Hi All,

We've been discussing fervently today internally around this, and while we've figured out a way to make this permission optional in chromium based browsers, obviously we won't be able to do so in Firefox.

After deliberation and discussion, and before our official product release announcement, we've decided that it would be better to exclude Firefox from browser biometric authentication, for now, until the upstream issue is resolved: https://bugzilla.mozilla.org/show_bug.cgi?id=1630415 rather than forcing all Firefox Bitwarden users to accept the new permission.

Extension update will be published soon as we're working on appropriate PRs to make this change, along with supporting documentation.

Thank you for your feedback and continued support, patience and input, it's extremely valuable and part of what makes open source amazing!

Sincerely,
The Bitwarden Team.

**Comment by ruifung** [comment]
> 
> 
> Hi All,
> 
> We've been discussing fervently today internally around this, and while we've figured out a way to make this permission optional in chromium based browsers, obviously we won't be able to do so in Firefox.
> 
> After deliberation and discussion, and before our official product release announcement, we've decided that it would be better to exclude Firefox from browser biometric authentication, for now, until the upstream issue is resolved: https://bugzilla.mozilla.org/show_bug.cgi?id=1630415 rather than forcing all Firefox Bitwarden users to accept the new permission.
> 
> Extension update will be published soon as we're working on appropriate PRs to make this change, along with supporting documentation.
> 
> Thank you for your feedback and continued support, patience and input, it's extremely valuable and part of what makes open source amazing!
> 
> Sincerely,
> The Bitwarden Team.

I do hope the feature is re-enabled soon as having it disabled just because a browser doesn't allow the permission to be optional is... sort of irritating for me. Especially since I use windows hello unlock.

Even more so since the version history on the mozilla addons page doesn't even state that its been delibrately removed... while the previous version says it was added. I guess I'll be sticking with 1.48.0 until such time where it's re-enabled.

Either that or I'll have to compile my own copy of it with the commit reverted.

**Comment by escape0707** [comment]
@ruifung Calm down, bro. Firefox side has already worked on it. https://phabricator.services.mozilla.com/D102998

**Comment by HVR88** [comment]
It would have been nice (and pertinent) to have included a note on the app side that browser integration was completely non-functional in Firefox. Browser integration is 99.9% of the interaction most people will ever have with BW.

**Comment by HVR88** [comment]
> @ruifung Calm down, bro. Firefox side has already worked on it. https://phabricator.services.mozilla.com/D102998

That's nice, but what that poster was noting is a Bitwarden issue - a Bitwarden marketing/engineering decision. The wrong decision.

**Comment by ruifung** [comment]
> 
> 
> @ruifung Calm down, bro. Firefox side has already worked on it. https://phabricator.services.mozilla.com/D102998

It's still going to be a while before the changes there get accepted (bugzilla says it's not yet) and for it to make it's way into a release version. 

I can understand where the bitwarden team is coming from, but having the function just suddenly get disabled because of what (to me personally) feels like a very minor thing is still somewhat irritating. And I had to go looking to even know it was disabled.

In the meantime, I've compiled my own copy with the commit reverted so I can keep using the feature.

>
> It would have been nice (and pertinent) to have included a note on the app side that browser integration was completely 
> non-functional in Firefox. Browser integration is 99.9% of the interaction most people will ever have with BW.

From my point of view, it's not non-functional... it's been delibrately disabled because the native messaging permission couldn't be made optional after people got concerned over the native messaging permisison. As someone who actually uses it for windows hello face unlock... having it suddenly just wait indefinitely at the "waiting for desktop application" message one day after it was clearly working without issue beforehand... yeah I got a bit annoyed there when I had to go looking only to find out it was disabled for what felt like a pretty silly reason to me.

**Comment by jaredm4** [comment]
Stop, @ruifung . Just because you're the one person who actually wants or apparently NEEDS this feature right now doesnt mean the backlash isn't justified. Even BW agreed they botched this release. Let it go.

It will come back. You lived with it prior to this week, you'll live a little longer.

**Comment by ruifung** [comment]
> 
> 
> Stop, @ruifung . Just because you're the one person who actually wants or apparently NEEDS this feature right now doesnt mean the backlash isn't justified. Even BW agreed they botched this release. Let it go.
> 
> It will come back. You lived with it prior to this week, you'll live a little longer.

Uh what...? At this point I'm really just responding to responses to my initial comment. I already solved my immediate want for this feature right now by well... compiling the plugin from source to re-enable it. I'm just stating my opinions and how I feel on the matter. Everything I've said is purely how I feel about it and my opinions on it (and the backlash, which yes, I disagree with, but that's beside the point).

I am under no delusions that anything I say here is going to change anything immediately, especially since it's after the fact, but all I've done is state my opinions and feelings on the matter. And I thought  I was being quite non-confrontational about the whole thing too.

That being said, I don't have to live without it. I can live with having to re-compile the extension everytime they update it for myself. That is one of the benefits of open source yes? If you don't like something personally, you can change it yourself. At least for your own copy anyway.

**Comment by abhigyank** [comment]
(not sure if this is just me or everyone still putting it out there if someone else faces the same problem) -
When the new permission was added and the extension was disabled, I didn't give it permission and it remained disabled. When the new version 1.48.1 was released, the extension auto-updated but it still kept asking for this permission even though the release said it was an optional permission.  I would have expected the auto-updated extension to not ask the permission once on the latest version with the permission optional. I had to remove the extension and reinstall it to not have this permission being asked and work as before. (this is on Chrome and Brave )

**Comment by rodalpho** [comment]
I also disagree with this decision. Mozilla takes _forever_ to make changes, even with a clean PR provided. I will honestly be surprised if this bug fix makes it to a released Firefox version in the year 2021.

**Comment by corvec** [comment]
Just chiming in to say that I also disagree with the removal of this feature. Biometric unlock in the browser was the one that convinced me to start switching from a competitor to Bitwarden. Now I need to figure out if it's worth it to me to manually compile the extension every time there's an update or if I should just hold off on switching.

**Comment by erasma** [comment]
This should not be difficult to understand no one is saying remove the
feature we saying give us the option to turn it off and allow us to be in
control of what we share and secure

**Comment by rodalpho** [comment]
Yes, but unfortunately that isn't possible unless BW either releases a separate copy of the extension or waits until Mozilla allows that permission to be marked optional. So they chose to remove the feature from everybody to avoid the scary permission dialog. 

If Mozilla really fixes the problem in a month or two that's no biggie. Knowing how Mozilla operates that is possible, but it's not likely.

**Comment by omnitec** [comment]
On Mon, 1 Feb 2021, rodalpho wrote:

> Yes, but unfortunately that isn't possible unless BW either releases a separate copy of the extension or waits until Mozilla allows that permission to be marked optional. So they chose to remove the feature from everybody to avoid the scary permission dialog.
>
Worked fine yesterday, .. all I had to do is select the "Authorize"
option?

**Comment by macav** [comment]
It seems that the issue has been fixed by Mozilla (https://bugzilla.mozilla.org/show_bug.cgi?id=1630415). Not sure when this is eventually released, but does it mean we can expect to have biometrics working in Firefox soon?

**Comment by rodalpho** [comment]
It's stalled waiting on documentation now, but if it passes review in time it'll be in Fx 87, due to be released march 23 2021. Assumedly once the GA Fx release supports that optional permission BW will quickly update their addon to suit.

**Comment by vhschlenker** [comment]
As per their [blog post](https://blog.nightly.mozilla.org/2021/02/26/these-weeks-in-firefox-issue-88/) under "WebExtension APIs":
> Starting from Firefox 87 nativeMessaging can be used as an optional permission
